### ISSUE TITLE
Nodes sometimes run out of internal IPs

### ISSUE THREAD

**Issue by pboothe** [m-lab/k8s-support#342]
Nodes in sandbox are sometimes rendered useless because their internal IP addresses are somehow registered as all used up.

We don't see this bug as much in staging or in prod because (I think) that the prod and staging machines have a very different ratio of deployments to reboots (they reboot more often). If I am right, then the bug is related to how we hand back (or not!) IP addresses after a pod is killed.

**Comment by pboothe** [comment]
Consensus is that there is an internal IP address assignment leak.

**Comment by nkinkade** [comment]
This problem still exists.  Using shimcni in sandbox on mlab2-lga0t, I discovered this:

```
==CMD==
/opt/shimcni/bin/multus
==ENV==
CNI_CONTAINERID=41a1947e91053c86553e06609639daf67e368504e48f75abb5c6a39262a65c8d
PWD=/
CNI_IFNAME=eth0
CNI_NETNS=/var/run/netns/cni-782bc44b-4527-0ea1-34b8-de148d5504cd
LANG=C.UTF-8
INVOCATION_ID=0cf7a4b79e4546b8b2e2acc8e9066aa3
CNI_COMMAND=DEL
SHLVL=1
CNI_PATH=/opt/shimcni/bin
LD_LIBRARY_PATH=/opt/containerd/lib:
CNI_ARGS=K8S_POD_NAME=ndt-go116-canary-p8fmb;K8S_POD_INFRA_CONTAINER_ID=41a1947e91053c86553e06609639daf67e368504e48f75abb5c6a39262a65c8d;IgnoreUnknown=1;K8S_POD_NAMESPACE=default
JOURNAL_STREAM=9:193140635
PATH=/opt/containerd/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
_=/usr/bin/env
==STDIN==
{"clusterNetwork":"flannel-conf","cniVersion":"0.3.1","kubeconfig":"/etc/kubernetes/kubelet.conf","multusNamespace":"default","name":"multus-network","type":"multus"}
==STDOUT==
{
    "code": 100,
    "msg": "Multus: error in invoke Delegate del - \"\": no plugin name provided"
}
==STDERR==
==EXITCODE==
1
```

This could explain the IP leak, though I do not yet know what is causing that error.

**Comment by nkinkade** [comment]
Investigating further, it would appear that the multus error above is not the cause, but I think I may have finally stumbled across a plausible explanation. The `host-local` IPAM stores IP allocation data, by default, at `/var/lib/cni/networks/<name>`. The `cmdDel()` function ultimately walks this data directory:

https://github.com/containernetworking/plugins/blob/master/plugins/ipam/host-local/backend/disk/backend.go#L140

The [documentation for `filepath.Walk()`](https://pkg.go.dev/path/filepath#Walk) states:

> Walk does not follow symbolic links.

However, what we have in the filesystem is:

```
root@mlab2-lga0t:/tmp# ls -l /var/lib/cni/networks/
total 0
drwxr-xr-x 2 root root 360 Sep 13 22:16 flannel
lrwxrwxrwx 1 root root   7 Sep 13 20:52 flannel-conf -> flannel
lrwxrwxrwx 1 root root   7 Sep 13 20:52 flannel-experiment-conf -> flannel
```

We have this unusual symlink configuration because the platform cluster had two separate cluster networks, one for experiments and one for non-experiments. The only difference between the two is the default gateway. In experiment pods the default gateway is the ipvlan interface ("net1") and in non-experiment pods the default gateway is bridged interface "eth0".  With two separate experiment directories, the `host-local` CNI plugin will store state in different locations, meaning that it could double assign a particular cluster IP on a given node. The symlinks cause the host-local plugin to effectively store all state in a single directory for both networks. 

It appears that the [`host-local` plugin does support a `dataDir` option](https://www.cni.dev/plugins/current/ipam/host-local/). Perhaps we can define this option in both cluster network configurations such that state will be stored in a single place regardless of which network is being configured, obviating the need for the awkward symlink setup. This option might be able to be specified in the `delegates` section of the `flannel` CNI configuration, causing it be passed through to the `host-local` plugin:

https://github.com/m-lab/k8s-support/blob/master/k8s/networks/networks.jsonnet#L13

**Comment by nkinkade** [comment]
Strike the idea of plumbing through the `dataDir` field for the `host-local` plugin. That field does not let you set the directory where state is held, but only the parent directory:

https://github.com/containernetworking/plugins/blob/master/plugins/ipam/host-local/backend/disk/backend.go#L47

**Comment by nkinkade** [comment]
The solution ended up being really trivial, and _should_ be resolved by #615. The issue was that we were not defining the `name` field in the CNI config for the flannel Network Attachment Definitions (NADs). When that field is missing in the CNI config, Multus apparently populates it using the `name` field of the NAD, effectively creating two separate networks.

**Comment by nkinkade** [comment]
The fix for this issue has now been deployed to production, and I believe this can be considered resolved.
